### PR TITLE
Add attributes to allowlist which can be accessed through objects (closes #5031 and #5032 )

### LINKED ISSUE #5031: TODOs fix - Torchvision - mean and std
## Description
In `src/syft/lib/torchvision/allowlist.py` we have a TODO: `mean and std are actually tuples`. 

## Additional Context
The fix has to be made and a test has to be written.

## Note
This issue will not be assigned to anybody until a PR is opened.

### LINKED ISSUE #5032: TODOs fix - Torchvision - normalize
## Description
In `src/syft/lib/torchvision/allowlist.py` we have a TODO: ` Normalize properties only exists on the object not on the class?`. 

## Additional Context
The fix has to be made and a test has to be written that the methods are working and are added to the allowlist.

## Note
This issue will not be assigned to anybody until a PR is opened.

### PR DESCRIPTION
**Pull request by avinsit123**
## Description

Closes #5031 
Closes #5032 

Currently, the problem with adding attributes in `allowlist` dictionary is that only those attributes can be added which are static and only accessible by class. This is not possible  if you will try to access any attribute of `torchvision.transforms.Normalize` like `inplace`  or `mean`. You get an error as shown below:-

These attributes can only be accessed when we pass the required positional arguments list as given below. The trick is to pass these positional arguments whenever we call the attributes like `mean` and `std`

### Fix
My proposed fix adds two new attribute to `src/syft/ast/attribute.py`  -  `self.require_pargs` and `self.parg_list` .
`self.require_pargs` is set to `True` is that attribute requires any positional arguments to access its sub-attributes, otherwise by default it is set to `False `. `self.parg_list` contains list of all the positional arguments that need to be passed into the attribute so that we can access its sub-attributes, by Default it is empty. As all components of AST like `Module` , `Class` , `Callable`  inherit from `attribute` class they inherit these 2 parameters also.

We check if `self.require_pargs` is set to `True` and pass `self.parg_list` into the class as shown in the code so that we can access its internal sub-attributes during using `getattr()`.

### How to use
Specified in the code. In case of `torchvision.transforms.Normalize` we need to pass in 2 attributes `mean` and `std` . Add name of attribute( in this case `torchvision.transforms.Normalize`) and list of positional arguments as corresponding entry in this case `[(1, 2, 3), (1, 2, 3)]` . Have a look at `src/syft/lib/torchvision/allowlist.py ` . 

## Affected Dependencies
Added new dict module_pargs in which modules that require positional arguments to access attributes can list their positional arguments.

## How has this been tested?
Wish to get some guidance for added fix and then write tests

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

**Comment by madhavajay** [comment]
@avinsit123 Looks like its going to be best to cherry pick these back off the updated `dev` branch. Sorry about this we had a really long standing branch with `0.4` and now its rebased onto `dev`.

I can do this if you would prefer when I get a moment.

**Comment by avinsit123** [comment]
I had closed my previous PR solving the same issue as @tudorcebere had asked me to rebase it with 0.4 : ) That's fine. It would be great if u could do it if u have time, otherwise I will do it. No issues.

**Comment by avinsit123** [comment]
@madhavajay Have a look at PR #5117 I have cherry-picked commits and rebased it with dev branch. Also could u help me out in how to write allowlist tests. Thanks.

**Comment by madhavajay** [comment]
Hi @avinsit123 Thanks for your PR. Sorry for the wait, I just went through this and I think I understand whats going on. In the event of p_args for a module_pargs key being defined, when the attribute is fetched an object is instantiated first and the attribute is fetched from the object instead of the class.

However, I created a simple test case and I can't seem to see any of the methods on the Pointer class.

```python
def test_object_properties() -> None:
    bob = sy.VirtualMachine(name="Bob")
    root_client = bob.get_root_client()

    ptr = root_client.torchvision.transforms.Normalize(1, 1)
    print(type(ptr))
    print("ptr inplace", getattr(ptr, "inplace", None))
    print("ptr mean", getattr(ptr, "mean", None))
    print("ptr std", getattr(ptr, "std", None))
    print(dir(ptr))
    assert hasattr(ptr, "inplace")
    assert hasattr(ptr, "mean")
    assert hasattr(ptr, "std")
```

Are you able to figure out why the Pointer is not getting these methods and perhaps add some tests for all the different scenarios that you have added code for so we can be sure they pass?

**Comment by avinsit123** [comment]
Hi, I see what might be happening before. It seems that attributes of Normalize like `mean` , `std` and `inplace`  are not getting added to the torchvision ast.

```python
>>> from syft.lib.torchvision import create_torchvision_ast
>>> create_torchvision_ast()
Module:
	.torchvision -> Module:
		.transforms -> Module:
			.Compose -> <syft.ast.klass.Class object at 0x7fac6130ed08>
			.ToTensor -> <syft.ast.klass.Class object at 0x7fac6130ed68>
			.Normalize -> <syft.ast.klass.Class object at 0x7fac6130edc8>

		.datasets -> Module:
			.MNIST -> <syft.ast.klass.Class object at 0x7fac61366048>
			.VisionDataset -> <syft.ast.klass.Class object at 0x7fac61366108>
```
This seems to happen with  `torchvision.datasets.MNIST.__len__` and  `torchvision.datasets.VisionDataset.__len__` both which were already present in allowlist before these changes. It seems that attributes of modules are not getting added to AST . I will try to fix this ASAP and write tests.

```python
>>> import syft as sy
>>> bob = sy.VirtualMachine(name="Bob")
>>> root_client = bob.get_root_client()
>>> ptr = root_client.torchvision.datasets.MNIST
>>> ptr.__len__
<syft.ast.callable.Callable object at 0x7ffeadbf2348>
[2021-02-08T21:34:12.205178+0530][CRITICAL][logger] __getattribute__ failed. If you are trying to access an EnumAttribute or a StaticAttribute, be sure they have been added to the AST. Falling back on__getattr__ to search in self.attrs for the requested field.
[2021-02-08T21:34:12.206367+0530][CRITICAL][logger] 'Class' object has no attribute '__len__'
>>> ptr.__len__
<syft.ast.callable.Callable object at 0x7ffeadbf2468>
[2021-02-08T21:34:56.009231+0530][CRITICAL][logger] __getattribute__ failed. If you are trying to access an EnumAttribute or a StaticAttribute, be sure they have been added to the AST. Falling back on__getattr__ to search in self.attrs for the requested field.
[2021-02-08T21:34:56.009740+0530][CRITICAL][logger] 'Class' object has no attribute '__len__'
```

**Comment by madhavajay** [comment]
@avinsit123 make sure to merge this into `dev` as the AST got a big rewrite and all that code has been merged into `dev` now.

**Comment by avinsit123** [comment]
Yup I have opened PR #5117 for this purpose

File: src/syft/lib/torchvision/allowlist.py [hunk 1]
```diff
@@ -12,11 +15,13 @@
 allowlist["torchvision.transforms.ToTensor"] = "torchvision.transforms.ToTensor"
 allowlist["torchvision.transforms.Normalize"] = "torchvision.transforms.Normalize"
 # TODO: Normalize properties only exists on the object not on the class?
-# allowlist["torchvision.transforms.Normalize.inplace"] = "syft.lib.python.Bool"
+allowlist["torchvision.transforms.Normalize.inplace"] = "syft.lib.python.Bool"
```
Comment (gmuraru) [inline src/syft/lib/torchvision/allowlist.py:18]: I am having a feeling that we might want the ```allowlist``` to become a more "standalone" class such that we might add more things to it? (like we have the ```module_pargs```)
What do you think @madhavajay, @tudorcebere ? or maybe I am overthinking this?

**Comment by madhavajay** [comment]
Closing this PR since this PR: https://github.com/OpenMined/PySyft/pull/5117 says it replaces it.

**Pull request closed without merging**